zk/conn.go

```diff
@@ -475,6 +487,17 @@ func (c *Conn) loop() {
 		}
 		c.flushRequests(err)
 
+		if err == ErrSessionExpired && !c.reconnectOnSessionExpiration {
+			atomic.StoreInt32(&c.sessionExpired, 1)
+
+			// The process of flushing requests is racy by design so do it in two steps.
+			c.flushUnsentRequests(err)
```

Comment:
This looks sloppy. Maybe it is better to handle expiration as an implicit call to Close(). That will drop requests that might be waiting for the lock. Alternatively, you can check the sessionExpired value before you do a request insert.